File: backend/src/api/routes/files.py
from fastapi import APIRouter, Depends, File, Form, UploadFile
from fastapi.responses import StreamingResponse

from src.api.deps import get_file_service
from src.schemas import FileItem, FileUpdate
from src.services.files import FileService
from src.tasks.pipeline import process_file

router = APIRouter(prefix="/files", tags=["files"])


@router.get("", response_model=list[FileItem])
async def list_files(service: FileService = Depends(get_file_service)):
    return await service.list()


@router.post("", response_model=FileItem, status_code=201)
async def create_file(
    title: str = Form(..., min_length=1, max_length=255),
    file: UploadFile = File(...),
    service: FileService = Depends(get_file_service),
):
    item = await service.create(title=title.strip(), upload_file=file)
    process_file.delay(item.id)
    return item


@router.get("/{file_id}", response_model=FileItem)
async def get_file(file_id: str, service: FileService = Depends(get_file_service)):
    return await service.get(file_id)


@router.patch("/{file_id}", response_model=FileItem)
async def update_file(
    file_id: str,
    payload: FileUpdate,
    service: FileService = Depends(get_file_service),
):
    return await service.update(file_id=file_id, title=payload.title.strip())


@router.get("/{file_id}/download")
async def download_file(file_id: str, service: FileService = Depends(get_file_service)):
    item, stream = await service.open_stream(file_id)
    return StreamingResponse(
        stream,
        media_type=item.mime_type,
        headers={
            "Content-Length": str(item.size),
            "Content-Disposition": f'attachment; filename="{item.original_name}"',
        },
    )


@router.delete("/{file_id}", status_code=204)
async def delete_file(file_id: str, service: FileService = Depends(get_file_service)):
    await service.delete(file_id)

File: frontend/src/lib/api/files.ts
import { apiFetch } from "@/lib/api/client";
import { apiBaseUrl } from "@/lib/config";
import type { FileItem } from "@/types/file";

export function listFiles(): Promise<FileItem[]> {
  return apiFetch<FileItem[]>("/files");
}

export function createFile(title: string, file: File): Promise<FileItem> {
  const form = new FormData();
  form.append("title", title);
  form.append("file", file);
  return apiFetch<FileItem>("/files", { method: "POST", body: form });
}

export function updateFile(id: string, title: string): Promise<FileItem> {
  return apiFetch<FileItem>(`/files/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ title }),
  });
}

export function deleteFile(id: string): Promise<void> {
  return apiFetch<void>(`/files/${id}`, { method: "DELETE" });
}

export function downloadUrl(id: string): string {
  return `${apiBaseUrl}/files/${id}/download`;
}
